packages/core/contracts/financial-templates/common/financial-product-libraries/PreExpirationIdentifierTransformationFinancialProductLibrary.sol

```diff
@@ -0,0 +1,59 @@
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+import "./FinancialProductLibrary.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+/**
+ * @title Pre-Expiration Identifier Transformation Financial Product Library
+ * @notice Adds custom identifier transformation to enable a financial contract to use two different identifiers, depending
+ * on when a price request is made. If the request is made before expiration than a transformation is made to the identifier
+ * & if it is at or after expiration then the original identifier is returned. This library enables self referential
+ * TWAP identifier to be used on synthetics pre-expiration, in conjunction with a separate identifier at expiration.
+ */
+contract PreExpirationIdentifierTransformationFinancialProductLibrary is FinancialProductLibrary, Ownable {
+    mapping(address => bytes32) financialProductTransformedIdentifiers;
+
+    /**
+     * @notice Enables the deployer of the library to set the transformed identifier for an associated financial product.
+     * @param financialProduct address of the financial product.
+     * @param transformedIdentifier the identifier for the financial product to be used if the contract is post expiration.
+     * @dev Note: a) Only the owner (deployer) of this library can set identifier transformations b) The identifier Cant
+     * be set to blank. c) A transformed price can only be set once to prevent the deployer from changing it after the fact.
+     * d)  financialProduct must exposes an expirationTimestamp method.
+     */
+    function setFinancialProductTransformedIdentifier(address financialProduct, bytes32 transformedIdentifier)
```

Comment:
Should this be `nonReentrant`?

Reply:
It is permissioned to be only called by the owner of the library and so reentrancy to is not really a risk here.

Reply:
Not sure I agree, the owner can be malicious and use a reentrancy attack after enough collateral has been deposited in the system. Also I think its better to just be safe and make every public method in the Perp system reentrancy-guarded

Reply:
no problem. I will add it 👍.